mlir/lib/Dialect/Rock/utility/transformMapUtils.cpp

```diff
@@ -346,9 +348,219 @@ propagateUnmergeVectorization(T &&dimAndLength,
+    SmallVector<size_t> dimPosition;
+    for (size_t j = i; j < upperDims.size(); j++) {
+
+      // Unit lengths don't affect the mergeability logic
+      if (params[j] == 1) {
+        continue;
+      }
+
+      auto keyJ = dimToMerge[upperDims[j]];
+
+      if (!keyJ.transform) {
+        break;
+      }
+      auto mergeJParams = keyJ.transform.getParams();
+      auto mergeJDims = keyJ.transform.getLowerDims();
+      size_t posJ = keyJ.positionInMerge;
+
+      // a) Dimensions need to come from the same merge
+      // b) Unmerge parameters need to match merge parameters
+      if (keyJ.transformPair() != keyI.transformPair() ||
+          params[j] != mergeJParams[posJ]) {
+        break;
+      }
+
+      groupCandidate.push_back(mergeJDims[posJ]);
+      dimPosition.push_back(posJ);
+    }
+
+    i += std::max(size_t(1), groupCandidate.size());
+
+    // Update the result with the current group for the mergePair key
+    if (groupCandidate.size() > 1 &&
+        std::is_sorted(dimPosition.begin(), dimPosition.end())) {
+
+      uint32_t lowerDim = groupCandidate.back();
+      for (auto d : groupCandidate) {
+        contiguousGroups[keyI.transformPair()].unionSets(lowerDim, d);
+      }
+    }
+  }
+}
+
+// This is an analysis pass to group the lower dimensions of a Merge
+// transformation into contiguous groups. E.g., if we have a Merge{8,8,3} [0] ->
+// [0,2,3] and we know that [2,3] are contiguous in the final representation, we
+// can split the merge group in
+// [[0] [2,3]]. This information will be used by the vectorizer. E.g., the
+// vectorizer can fulfill a vectorization by 4, since 8*3=24 is a multiple of 4.
+// In other words, every group of dimensions is treated as a single group
+ContiguousMergesMap findContiguousGroups(ArrayAttr transforms,
+                                         ArrayRef<int64_t> outputShape) {
+  // Transform table. Will be overwritten after processing each transform_map
+  DimToMergeMap dimToMerge;
+  ContiguousMergesMap contiguousGroups;
+
+  for (TransformMapAttr transformMap :
+       transforms.getAsRange<TransformMapAttr>()) {
+    ArrayRef<int64_t> upperBounds = transformMap.getUpperBounds();
+
+    DimToMergeMap thisDimToMerge;
+
+    for (TransformAttr transform : transformMap.getOps()) {
+      TransformType transformType = transform.getType();
+
+      ArrayRef<uint32_t> lowerDims = transform.getLowerDims();
+      ArrayRef<uint32_t> upperDims = transform.getUpperDims();
+      ArrayRef<int64_t> params = transform.getParams();
+
+      switch (transformType) {
+      case TransformType::Unfold:
+      case TransformType::Merge:
+        for (size_t i = 0; i < lowerDims.size(); i++) {
+          thisDimToMerge[lowerDims[i]] = {transformMap, transform, i};
+        }
+        break;
+      case TransformType::AddDim:
+        break;
+      case TransformType::Embed: {
+        // TODO: we don't support stacked embed/embed. So we want to make sure
```

Comment:
Can I ask what this case is about? I don't see the problem here - what complicates
```
(a, b) -> (a / 3, a % 3, b / 4, b % 4)
(a, b, c, d) -> (b + 3 * a, c + 4 * d)
```
from concluding that `a` and `b` get merged back contiguously?

Reply:
This was only because I didn't add any test for this case. I did just now (and fixed a small issue with the upper bounds).